sql, distsql: aggregation planning

irfansharif opened the pull request:
Introduce distributed sql physical planning for aggregation (`GROUP BY`).
Note that we currently do not support aggregation in presence of
`HAVING`, for e.g. `SELECT 3 FROM kv HAVING TRUE`, or the `ARRAY_AGG`
function due to the lack of support of array types in distsql.

Concisely we replace the `groupNode` in our `planNode` tree by adding
`aggregator` processors, an `evaluator` too if necessary.

See individual commits.
cc @knz @andreimatei

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12235)
<!-- Reviewable:end -->

irfansharif committed:
sql: miscellaneous cleanup

- Changed `checkExpr` to be a method attached to the `distSQLPlanner`
instead of a standalone function.
- Mirrored "wrap" logic in `SelectTopNode` when adding `sorters` to
accommodate another `planNode` to be it's "source".
- Added tests to check setting of DIST_SQL session variables.

irfansharif committed:
sql: fix selectRenders for non-mapped processors

The current `selectRenders` implementation does not accommodate for
processors that don't have internal "mappings", i.e. given input rows
with the schema `[k]` the processor itself can generate rows of the
schema `[k k k]`. One such example of this is the `aggregator`.

In cases such as this we will need to add an intermediary evaluator to
do the right multiplexing/column re-mapping as needed by the processor.

irfansharif committed:
sql: distributed aggregation planning

Introduce distributed sql physical planning for aggregation
(`GROUP BY`).
Note that we currently do not support aggregation in presence of
`HAVING`, for e.g. `SELECT 3 FROM kv HAVING TRUE`, or the `ARRAY_AGG`
function due to the lack of support of array types in distsql.

Concisely we replace the groupNode in our planNode tree by adding an
aggregator processor. We also add an evaluator in the case of
operations where the operands are themselves the results of aggregation
functions, like in "SUM(k) + SUM(v)".

RaduBerinde commented:
Thanks for working on this!!

I haven't yet looked in detail at the expression extraction stuff.

For the third commit message, can you explain at a high-level how the plan looks like (that we serialize everything into a single processor). I would also remove `distributed` from the title since we're not doing distributed aggregation yet, maybe just `distsql: aggregation planning`).

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 576 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaHVeH8Z6S7z-zJOrq:-KYaHVeH8Z6S7z-zJOrr:brnxhbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L576)):*
> ```Go
> // add an evaluator stage. The specs for such processors can definitely be
> // improved upon.
> func (dsp *distSQLPlanner) selectRenders(
> ```

This change is causing us to add evaluators where they are not needed (in cases that work fine now).. Plus the code would need to come back once the TODO above is fixed.. We should only be strict when it's needed - we could just check if `n.top.group != nil` and do the stricter check in that case.

---

*[pkg/sql/distsql_physical_planner.go, line 744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaI5VVy8QjcTFQLHVw:-KYaI5VVy8QjcTFQLHVx:bnx7inz) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L744)):*
> ```Go
> 
> 	case *selectTopNode:
> 		n.plan = n.source
> ```

Why are we modifying the node here? DistSQL planning should not modify anything in any `planNodes`.

---

*[pkg/sql/distsql_physical_planner.go, line 294 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaIjr2C2pCRPK00f5l:-KYaIjr2C2pCRPK00f5m:b-bkqfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L294)):*
> ```Go
> 	}
> 
> 	// TODO(irfansharif): currently there’s no nice way to “compose” FmtFlags
> ```

I think we should just make the flags public, we are unnecessarily making things harder for ourselves. The wrappers always seemed like unnecessary boilerplate to me. @knz what do you think?

---

*[pkg/sql/distsql_physical_planner.go, line 904 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaKEzy7j19ETa_GyIs:-KYaKEzy7j19ETa_GyIt:b-45mrnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L904)):*
> ```Go
> 		GroupCols: groupCols,
> 	}
> 	dsp.addSingleGroupStage(
> ```

Please put two TODOs here: 
 - we could add a local "no grouping" stage to pre-aggregate results before the final aggregation
 - we could route by hash to multiple final aggregators

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-:-KYaIRddkngB-90kVVV-:b-1rnbpj)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
One question I have is: does a query like `select sum(k + v) from kv group by k,v` work? Which visitor extracts the `k + v` expression and adds an evaluator for it?

I think what would improve my understanding of everything would be a description of the whole process, tying together the comments you now have on the pre and post aggregation visitors and also going into how the renderers for the source and the parent of the group node are generated. Perhaps move these visitors to a separate file and add a top-level comment there?

---

Review status: 0 of 3 files reviewed at latest revision, 24 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYo9CO7LvyAOQ7kHelY:-KYo9CO8hySSqNVANb9e:bgngzs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7099fb2bfab05d96d11579f996ca560ed97f59de/pkg/sql/distsql_physical_planner.go#L623)):*
> ```Go
> 		OutputOrdering: dsp.convertOrdering(n.ordering, p.planToStreamColMap),
> 		OrderingMatchLen: uint32(computeOrderingMatch(
> 			n.ordering, n.plan.Ordering(), false, /* reverse */
> ```

`addSorters` was being called with `addSorters(..., selectTopNode, selectTopNode.source)`. Is the last arg the same as `sortNode.plan`?
In any case, this call is too dense. I think it'd benefit from a comment explaining what match we're computing - between what and what.  

---

*[pkg/sql/distsql_physical_planner.go, line 567 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoELWChTvaWb7zFgLr:-KYoELWChTvaWb7zFgLs:b-im0yae) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L567)):*
> ```Go
> // references, or there isn't a strict 1-1 mapping between n.Columns() and
> // n.source.plan.Columns().
> // TODO(irfansharif): Arguably simple multiplexing or column re-mapping is too
> ```

I think it'd help to clarify this todo: add support <processors> and their specs for ordinal column references?

What else was currently broken, other than aggregation?

---

*[pkg/sql/distsql_physical_planner.go, line 744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaI5VVy8QjcTFQLHVw:-KYoB-u8iTzG2qf0Z60t:b-4iyxmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L744)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why are we modifying the node here? DistSQL planning should not modify anything in any `planNodes`.
</blockquote></details>

+1. What's going on here?

---

*[pkg/sql/distsql_physical_planner.go, line 286 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoHDn8Hnk9el04zgu8:-KYoHDn8Hnk9el04zgu9:bc15jlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L286)):*
> ```Go
> 
> // The distsql Expression uses the placeholder syntax (@1, @2, @3..) to
> // refer to columns, we format the expression using the IndexedVar and StarDatum
> ```

You've changed the period to a comma, but the period was better :)

---

*[pkg/sql/distsql_physical_planner.go, line 300 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoGh8K_yBbj5rbNQhq:-KYoGh8K_yBbj5rbNQhr:b-wpctt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L300)):*
> ```Go
> 	// the flags would not need to inspect expression type.
> 	var f parser.FmtFlags
> 	switch expr.(type) {
> ```

I don't quite understand what's going on. What if the `*` is just one component of `Expr`, not the whole expression? Don't we need to "compose" the start formatter with the other one (like you say in the TODO)?

---

*[pkg/sql/distsql_physical_planner.go, line 303 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoGT_de56A_K3pOsLF:-KYoGT_einm-LiLYVmoT:bm9wp22) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L303)):*
> ```Go
> 	case *parser.StarDatum:
> 		// By default parser.StarDatum is rendered as a DInt(0), but formatting
> 		// this as 0 we can replicate this behavior in distsql.
> ```

sorry, I don't understand this phrase. "we can replicate this behavior" - which behavior?

---

*[pkg/sql/distsql_physical_planner.go, line 710 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoHXB888ovZAUXR4bX:-KYoHXB888ovZAUXR4bY:b-vap8pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L710)):*
> ```Go
> }
> 
> type preAggExprVisitor struct {
> ```

please add a comment to this type about what the visitor does

---

*[pkg/sql/distsql_physical_planner.go, line 726 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoKLCe1Nd_n4Y6bSjU:-KYoKLCe1Nd_n4Y6bSjV:bbt2xcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L726)):*
> ```Go
> 
> 	fholder := expr.(*aggregateFuncHolder)
> 	if _, ok := fholder.expr.(*parser.FuncExpr); !ok {
> ```

mind adding a comment to `aggregateFuncHolder.expr` about what it can contain? If it's not an aggregate function call, then... it needs to be the same expression as one of the `GROUP BY` expressions?

---

*[pkg/sql/distsql_physical_planner.go, line 734 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoKleD0LltrH-WulxX:-KYoKleD0LltrH-WulxY:bk6z0tn) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L734)):*
> ```Go
> 	}
> 
> 	f := fholder.expr.(*parser.FuncExpr)
> ```

nit: assign this `f` on the first cast line, above.

---

*[pkg/sql/distsql_physical_planner.go, line 758 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoLoldxIqgG0SlQK-z:-KYoLoldxIqgG0SlQK0-:b-7mkh3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L758)):*
> ```Go
> // extractPreAggExprs translates the render expressions into the form needed by
> // the AggregatorSpec where per "aggregation function" we specify the function
> // type (`SUM`, `COUNT`, `AVG`, etc) and the column this function is going to be
> ```

probably worth to name IDENT as one of the function examples

---

*[pkg/sql/distsql_physical_planner.go, line 760 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoLz7Lsiry4qzCb4uy:-KYoLz7Lsiry4qzCb4uz:b-5hk849) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L760)):*
> ```Go
> // type (`SUM`, `COUNT`, `AVG`, etc) and the column this function is going to be
> // operating on (@1, @2, @1, etc).
> // In order to this we need to "flatten" out the render expressions, the
> ```

nit: several places in this PR you've used commas to separate sentences where that's not quite grammatically correct because the sentences are independent clauses and you're not using coordinating conjunctions. Or something :) At least it doesn't read correctly to me.
Here, for example, you want either a period or a semi-colon before "the following examples..."

---

*[pkg/sql/distsql_physical_planner.go, line 762 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoNosyoiAB5pLETJ8c:-KYoNoszSUX4bSOAfCW-:b-r0cktz) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L762)):*
> ```Go
> // In order to this we need to "flatten" out the render expressions, the
> // following examples detail out what we mean by this.
> // - `SELECT COUNT(k), SUM(v + w), v + w FROM kv GROUP BY v + w`
> ```

I think this example doesn't compute - `COUNT(k)` operates on a column that doesn't appear in `GROUP BY`. No?

---

*[pkg/sql/distsql_physical_planner.go, line 766 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoOgvYB-X7jc1GsKFs:-KYoOgvZ3Tuf2_t3rEot:b55izqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L766)):*
> ```Go
> //   expressions we are concerned with are [k v+w v+w]. We see that for `COUNT`
> //   the stream it needs to be fed is the first, for `SUM` it is the second, for
> //   `v + ` we add an `IDENT` aggregation function.
> ```

for v + **w**  we add ...

---

*[pkg/sql/distsql_physical_planner.go, line 769 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoPNyrE7NGlkTxaDfg:-KYoPNyrE7NGlkTxaDfh:b-23kq2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L769)):*
> ```Go
> //   Already we see that for the n-th aggregation function we operate on the
> //   n-th stream.
> // - `SELECT COUNT(k) + COUNT(v) FROM kv`
> ```

again I think you need a GROUP BY

---

*[pkg/sql/distsql_physical_planner.go, line 774 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoPRWGUOlXMD0o-kdV:-KYoPRWGUOlXMD0o-kdW:bqnhow2) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L774)):*
> ```Go
> //   to operate on the first stream, another `COUNT` to operate on the second.
> //   Again we see that for the n-th aggregation function we operate on the n-th
> //   stream (note, this is distinct from the n-th render expression).
> ```

if you're gonna mention the render expressions, maybe mentions who's gonna finally compute them - an evaluator introduced by the select top node?

---

*[pkg/sql/distsql_physical_planner.go, line 780 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoPlbOPqVHuFWd6gWP:-KYoPlbOPqVHuFWd6gWQ:bepb2at) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L780)):*
> ```Go
> 	v := preAggExprVisitor{}
> 	for _, expr := range render {
> 		aggExprs = append(aggExprs, v.extract(expr)...)
> ```

should there be a TODO somewhere to dedupe `aggExprs` ?

---

*[pkg/sql/distsql_physical_planner.go, line 788 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoQGsbiGctIjQvpWRr:-KYoQGsbiGctIjQvpWRs:bkzfkwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L788)):*
> ```Go
> }
> 
> type postAggExprVisitor struct {
> ```

please add a comment explaining what this visitor does

---

*[pkg/sql/distsql_physical_planner.go, line 805 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoStnutN9CbOmmUQiC:-KYoStnutN9CbOmmUQiD:bfkbe01) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L805)):*
> ```Go
> // instead of VisitPre, an example to demonstrate why is that if we have
> // `SELECT COUNT(k) + COUNT(v) FROM kv`, for our render expression
> // `COUNT(k) + COUNT(v)` we want it to be substituted by `@1 + @2`. Therefore we
> ```

I don't understand what this example is trying to tell me :) Why is the evaluation order between the parent and the children important? It's no possible for both parent and child to increment `v.count`, is it?

---

*[pkg/sql/distsql_physical_planner.go, line 882 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoYSdjnm4X-KW15IsJ:-KYoYSdjnm4X-KW15IsK:b-ry8vns) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L882)):*
> ```Go
> //      - Expressions that CONTAIN an aggregation function, e.g. `COUNT(k) + 1`.
> //        This is evaluated the post aggregation evaluator attached after.
> //    All other expressions simply pass through unchanged, for e.g. `1` in
> ```

what about `k` in `select k group by k` ? Mention it in this example?

---

*[pkg/sql/distsql_physical_planner.go, line 888 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoYbhKh6CiR-f5Jzet:-KYoYbhKh6CiR-f5Jzeu:byrygi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L888)):*
> ```Go
> 	types := dsp.getTypes(n.plan.Columns())
> 	// The way our planNode construction currently orders columns between
> 	// groupNode and it's source is that the first len(n.funcs) columns are the
> ```

s/it's/its

---

*[pkg/sql/testdata/set, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoBblbuAmDhOxKzVQK:-KYoBblcZ44h3nkuoMsa:bbracfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7099fb2bfab05d96d11579f996ca560ed97f59de/pkg/sql/testdata/set#L147)):*
> ```
> 
> statement ok
> SET DIST_SQL = OFF
> ```

might as well add a test from `set dist_sql = bogus`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-:-KYoRZqkMEbCbGD8W_o0:bq4biqe)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 31 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 710 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoHXB888ovZAUXR4bX:-KYyZkHVr4kyiJQBvY29:bk9rmhp) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L710)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment to this type about what the visitor does
</blockquote></details>

+1, also explain the name (`pre` vs `post` below)

---

*[pkg/sql/distsql_physical_planner.go, line 725 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYyZS8fuRfqw88b8kqV:-KYyZS8fuRfqw88b8kqW:bfu8g07) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L725)):*
> ```Go
> 	}
> 
> 	fholder := expr.(*aggregateFuncHolder)
> ```

why are we doing this casting twice? it's not that important to keep `ok` in its own scope.. Same below.

---

*[pkg/sql/distsql_physical_planner.go, line 752 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYyYlgMrLJDMXfV_Gwz:-KYyYlgMrLJDMXfV_Gx-:b505k80) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L752)):*
> ```Go
> 
> func (v preAggExprVisitor) extract(typedExpr parser.TypedExpr) []distsql.AggregatorSpec_Expr {
> 	parser.WalkExpr(&v, typedExpr)
> ```

WalkExprConst

---

*[pkg/sql/distsql_physical_planner.go, line 759 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYy_Sk5zVTeDFFVqWML:-KYy_Sk5zVTeDFFVqWMM:bxas1wp) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L759)):*
> ```Go
> // the AggregatorSpec where per "aggregation function" we specify the function
> // type (`SUM`, `COUNT`, `AVG`, etc) and the column this function is going to be
> // operating on (@1, @2, @1, etc).
> ```

@3 and "in order to do this" below

---

*[pkg/sql/distsql_physical_planner.go, line 762 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYy_cbZUc3CRQbNvmYY:-KYy_cbZUc3CRQbNvmYZ:bbous1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L762)):*
> ```Go
> // In order to this we need to "flatten" out the render expressions, the
> // following examples detail out what we mean by this.
> // - `SELECT COUNT(k), SUM(v + w), v + w FROM kv GROUP BY v + w`
> ```

markdown isn't useful in comments, i'd remove the backquotes

---

*[pkg/sql/distsql_physical_planner.go, line 762 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoNosyoiAB5pLETJ8c:-KYyaVai2ITLCEW_1VC5:b-4znd5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L762)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this example doesn't compute - `COUNT(k)` operates on a column that doesn't appear in `GROUP BY`. No?
</blockquote></details>

For each value of `v+m`, this is the number of rows with non-NULL `k`.

Actually the one I'm a bit (ra)dubious of semantically is `SUM(v + w)` - I think in this case it's always equal to `(v+w) * COUNT(*)`.  It's good enough for the purposes of this example though.

---

*[pkg/sql/distsql_physical_planner.go, line 763 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYy_nX9CcUdgDh6pCl-:-KYy_nX9CcUdgDh6pCl0:b-aykgjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L763)):*
> ```Go
> // following examples detail out what we mean by this.
> // - `SELECT COUNT(k), SUM(v + w), v + w FROM kv GROUP BY v + w`
> //   The output schema of groupNode's source here is [k v+w v+w v+w]`, the render
> ```

I would say "of selectNode (groupNode's source)". also, extra backquote

---

*[pkg/sql/distsql_physical_planner.go, line 764 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYybi4tvE1_jncU0JNx:-KYybi4tvE1_jncU0JNy:bj6w3mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L764)):*
> ```Go
> // - `SELECT COUNT(k), SUM(v + w), v + w FROM kv GROUP BY v + w`
> //   The output schema of groupNode's source here is [k v+w v+w v+w]`, the render
> //   expressions we are concerned with are [k v+w v+w]. We see that for `COUNT`
> ```

phrasing is very awkward

---

*[pkg/sql/distsql_physical_planner.go, line 769 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYyblczlGDE5PFqBAsZ:-KYyblczlGDE5PFqBAs_:brvujac) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L769)):*
> ```Go
> //   Already we see that for the n-th aggregation function we operate on the
> //   n-th stream.
> // - `SELECT COUNT(k) + COUNT(v) FROM kv`
> ```

[nit] add some blank lines between the examples

---

*[pkg/sql/testdata/set, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoBblbuAmDhOxKzVQK:-KYyYutkIP7-eK3R3XMW:bcovvcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7099fb2bfab05d96d11579f996ca560ed97f59de/pkg/sql/testdata/set#L147)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

might as well add a test from `set dist_sql = bogus`
</blockquote></details>

We should make `bogus` the current default.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
> does a query like select sum(k + v) from kv group by k,v work? Which visitor extracts the k + v expression and adds an evaluator for it?

Yes it works, and the answer is neither :) in this case the `k+v` evaluation actually happens at the `selectNode`. The assumption for `aggregators` is quite simple,  we should have our Spec simply state the LIST of aggregation functions, and it'll operate on pre-evaluated column streams (in this case a stream of already evaluated `k+v` values).

---

Review status: 0 of 5 files reviewed at latest revision, 31 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYo9CO7LvyAOQ7kHelY:-KYzTCkF6dqEk6iGSwq2:b-sqgvht) ([raw file](https://github.com/cockroachdb/cockroach/blob/7099fb2bfab05d96d11579f996ca560ed97f59de/pkg/sql/distsql_physical_planner.go#L623)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`addSorters` was being called with `addSorters(..., selectTopNode, selectTopNode.source)`. Is the last arg the same as `sortNode.plan`?
In any case, this call is too dense. I think it'd benefit from a comment explaining what match we're computing - between what and what.  
</blockquote></details>

the last argument wasn't *always* the same no, hence the change. if there's a `groupNode` *and* a `sortNode` (like `SELECT ... GROUP BY ... ORDER BY ...`), the `sourceNode` that had to be passed in would have to be the `groupNode`. By taking advantage of the `wrap` that happens in `select_top.go` we can instead refer to `n.plan` to be it's "source".

---

*[pkg/sql/distsql_physical_planner.go, line 567 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoELWChTvaWb7zFgLr:-KYzSq7M_29IO99g6L24:b7fj188) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L567)):*
Done. 

> What else was currently broken, other than aggregation?
sorting as well because of the lack of a "projection". Added comments to explain.

---

*[pkg/sql/distsql_physical_planner.go, line 576 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaHVeH8Z6S7z-zJOrq:-KYzT0i-ddrI91EfmzA_:bjyj0ag) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L576)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This change is causing us to add evaluators where they are not needed (in cases that work fine now).. Plus the code would need to come back once the TODO above is fixed.. We should only be strict when it's needed - we could just check if `n.top.group != nil` and do the stricter check in that case.
</blockquote></details>

Done. Note that `n.top.group != nil` is not sufficient, also needed a check for `n.top.sort != nil`. Reason added to the comments.

---

*[pkg/sql/distsql_physical_planner.go, line 744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaI5VVy8QjcTFQLHVw:-KYzVqZ5o_l2GpGboKB9:b-pj2lk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L744)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

+1. What's going on here?
</blockquote></details>

This was a mistake, I misunderstood something @raduberinde mentioned to me earlier but changed this.

---

*[pkg/sql/distsql_physical_planner.go, line 286 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoHDn8Hnk9el04zgu8:-KYzSiHB0LTkw6iL7zaA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L286)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

You've changed the period to a comma, but the period was better :)
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 294 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaIjr2C2pCRPK00f5l:-KYzWB2EAmcvOw45gESy:bvec9ki) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L294)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think we should just make the flags public, we are unnecessarily making things harder for ourselves. The wrappers always seemed like unnecessary boilerplate to me. @knz what do you think?
</blockquote></details>

probably a separate PR from this.

---

*[pkg/sql/distsql_physical_planner.go, line 300 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoGh8K_yBbj5rbNQhq:-KYzWGZnsHj9Uc1ZHvyx:bnpu1x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L300)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't quite understand what's going on. What if the `*` is just one component of `Expr`, not the whole expression? Don't we need to "compose" the start formatter with the other one (like you say in the TODO)?
</blockquote></details>

what's an example where * is just one component of `Expr`? I believe the only time this comes into play is when we have `COUNT(*)`.

---

*[pkg/sql/distsql_physical_planner.go, line 303 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoGT_de56A_K3pOsLF:-KYyqUvAAlnowqnn0CZo:bwxhaav) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L303)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

sorry, I don't understand this phrase. "we can replicate this behavior" - which behavior?
</blockquote></details>

For e.g. in `SELECT COUNT(*) from kv GROUP BY k`', despite the output schema for
selectNode being `[* k]`, when evaluated it generates `[0 k]` for every row.
		
By formatting this as '0' the `selectNode` equivalent in distsql will also produce 0's as needed.
Does that help?

---

*[pkg/sql/distsql_physical_planner.go, line 710 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoHXB888ovZAUXR4bX:-KYz_cnQAF0ZvbwN-cho:bexoxrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L710)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

+1, also explain the name (`pre` vs `post` below)
</blockquote></details>

hmm, I tried explaining the workings of this in the corresponding `.extract(...)` method. Would you like me to add something in *addition* to that?

---

*[pkg/sql/distsql_physical_planner.go, line 725 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYyZS8fuRfqw88b8kqV:-KYzU4EMyfWcUZYkqWBy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L725)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

why are we doing this casting twice? it's not that important to keep `ok` in its own scope.. Same below.
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 726 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoKLCe1Nd_n4Y6bSjU:-KYzU3ZoSLgAg8C0ddu3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L726)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

mind adding a comment to `aggregateFuncHolder.expr` about what it can contain? If it's not an aggregate function call, then... it needs to be the same expression as one of the `GROUP BY` expressions?
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 734 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoKleD0LltrH-WulxX:-KYzU52r1UVke-RrXYZl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L734)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: assign this `f` on the first cast line, above.
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 752 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYyYlgMrLJDMXfV_Gwz:-KYzU5qMzOagEP_qhGCv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L752)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

WalkExprConst
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 758 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoLoldxIqgG0SlQK-z:-KYzU6a9nOkvIHK9O7jy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L758)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

probably worth to name IDENT as one of the function examples
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 759 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYy_Sk5zVTeDFFVqWML:-KYzVZKrbspx_jhkOmA8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L759)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

@3 and "in order to do this" below
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 760 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoLz7Lsiry4qzCb4uy:-KYzU7IgZym4MkKpvv6W:b-gfl9jj) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L760)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: several places in this PR you've used commas to separate sentences where that's not quite grammatically correct because the sentences are independent clauses and you're not using coordinating conjunctions. Or something :) At least it doesn't read correctly to me.
Here, for example, you want either a period or a semi-colon before "the following examples..."
</blockquote></details>

Done, thank you for your explanation haha.

---

*[pkg/sql/distsql_physical_planner.go, line 762 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYy_cbZUc3CRQbNvmYY:-KYzVZrTSdhO18-1ThfO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L762)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

markdown isn't useful in comments, i'd remove the backquotes
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 762 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoNosyoiAB5pLETJ8c:-KYzYP8HVNrJKChTlJDC:b-w1xzf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L762)):*
> `SUM(v + w)` [...] always equal to `(v+w) * COUNT(*)`
Yes.

---

*[pkg/sql/distsql_physical_planner.go, line 763 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYy_nX9CcUdgDh6pCl-:-KYzVa6CD3UPMPxlJtIw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L763)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I would say "of selectNode (groupNode's source)". also, extra backquote
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 764 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYybi4tvE1_jncU0JNx:-KYzVag4dGKOS7g9dyYP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L764)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

phrasing is very awkward
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 766 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoOgvYB-X7jc1GsKFs:-KYzUFj-iSBhQeoBc27q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L766)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

for v + **w**  we add ...
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 769 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoPNyrE7NGlkTxaDfg:-KYzUe5Jla4YqQI-1Rh-:b-yjy5wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L769)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

again I think you need a GROUP BY
</blockquote></details>

`SELECT COUNT(k) + COUNT(v) FROM kv` no we don't :), here the `GROUP BY` is implicitly nothing so we compute `COUNT(k)`, `COUNT(v)` over the entire table.

---

*[pkg/sql/distsql_physical_planner.go, line 769 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYyblczlGDE5PFqBAsZ:-KYzVb8ZSThfX5dzDPsq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L769)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add some blank lines between the examples
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 774 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoPRWGUOlXMD0o-kdV:-KYzUrbidW_DOhz_2yrt:bgppvtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L774)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if you're gonna mention the render expressions, maybe mentions who's gonna finally compute them - an evaluator introduced by the select top node?
</blockquote></details>

Done. Actually the evaluator is added with the expressions extracted via `postAggrExprsVisitor` and is introduced following `evalExprs, needEval := dsp.extractPostAggrExprs(n.render)`  if `needEval == true` but added a comment explaining this :)

---

*[pkg/sql/distsql_physical_planner.go, line 780 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoPlbOPqVHuFWd6gWP:-KYzVRbsxvStiYX3ruip:b-lpuua9) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L780)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

should there be a TODO somewhere to dedupe `aggExprs` ?
</blockquote></details>

@knz, mentioned this to you earlier. Is there an issue already tracking this?

---

*[pkg/sql/distsql_physical_planner.go, line 788 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoQGsbiGctIjQvpWRr:-KYz_sOuggf02U6559zj:bwu9q8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L788)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment explaining what this visitor does
</blockquote></details>

look at the other related comment here, I added the explanation for the corresponding `.extract(...)` method explaining the working of this. Would you like some *additional* explanation?

---

*[pkg/sql/distsql_physical_planner.go, line 805 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoStnutN9CbOmmUQiC:-KYzYoiOkGKzxuuHkrEh:bugho73) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L805)):*
> It's not possible for both parent and child to increment v.count, is it?
You are correct, it's not possible.

> Why is the evaluation order between the parent and the children important?
Because we need the children of `+` to be `@1` and `@2`, hence before constructing the `+` we need to *first* produce `@1` and `@2`. Does that make it clearer? I admit it's a bit tough to explain this :)

---

*[pkg/sql/distsql_physical_planner.go, line 882 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoYSdjnm4X-KW15IsJ:-KYzVg_25xPIDInZZdME:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L882)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what about `k` in `select k group by k` ? Mention it in this example?
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 888 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoYbhKh6CiR-f5Jzet:-KYzVpClEqlt33gsTu3a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L888)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/it's/its
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 904 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaKEzy7j19ETa_GyIs:-KYzVpqIrTg46ndezeAi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L904)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Please put two TODOs here: 
 - we could add a local "no grouping" stage to pre-aggregate results before the final aggregation
 - we could route by hash to multiple final aggregators
</blockquote></details>

Done.

---

*[pkg/sql/testdata/set, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoBblbuAmDhOxKzVQK:-KYzW00a0lpLHRWw2d-b:b-ijr4oa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7099fb2bfab05d96d11579f996ca560ed97f59de/pkg/sql/testdata/set#L147)):*
> might as well add a test from set dist_sql = bogus
Done.

> We should make bogus the current default.
not sure I follow.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-:-KYzbYCEKqqn1-nL88O9:b9rbe0q)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
gentle ping :)
I can move over the two visitors to a separate file and move around the documentation to have some defined on the struct definition itself. As per Andrei's suggestion also add a top-level comment there explaining everything.

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 31 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 710 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYoHXB888ovZAUXR4bX:-KZ2wSdTMPvKhgKVQFMY:bvdloo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f06cb259e249fd89e77a4d3679235103f20cf4/pkg/sql/distsql_physical_planner.go#L710)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

hmm, I tried explaining the workings of this in the corresponding `.extract(...)` method. Would you like me to add something in *addition* to that?
</blockquote></details>

That's fine, can you just add a "see extractPreblabla" so the reader knows where to jump to for the explanation?

---

*[pkg/sql/distsql_physical_planner.go, line 685 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KZ2rkqPC5nCEstCWCcu:-KZ2rkqQITgBQIgpuyr1:b-8aczzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9cc2ec2c23dcfe45b2817ab5867ce5545c09657/pkg/sql/distsql_physical_planner.go#L685)):*
> ```Go
> // select node has any expressions which are not just simple column references.
> //
> // An evaluator stage is also added if there isn't a strict
> ```

is also added in some cases where a strict 1-1 mapping is required between the planNode columns and the distSQL streams.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-:-KZ2wjFO8OPK7HPiaAqg:bnfp4nl)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Moved over the visitors to a separate `distsql_aggregate_visitors.go`. thanks for the review!

---

Review status: 0 of 6 files reviewed at latest revision, 31 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsql_physical_planner.go, line 685 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KZ2rkqPC5nCEstCWCcu:-KZ7WU_Z43jQEXnHj7Di:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9cc2ec2c23dcfe45b2817ab5867ce5545c09657/pkg/sql/distsql_physical_planner.go#L685)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

is also added in some cases where a strict 1-1 mapping is required between the planNode columns and the distSQL streams.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-:-KZ7WOnAtFq0K5HUePQQ:b16sut5)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/distsql_physical_planner.go, line 576 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaHVeH8Z6S7z-zJOrq:-KZP0HS5hFpeNjrqHU7x:b98zd9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L576)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Done. Note that `n.top.group != nil` is not sufficient, also needed a check for `n.top.sort != nil`. Reason added to the comments.
</blockquote></details>

By the way - why is 1-1 mapping necessary for grouping? Can't we simply do `ColIdx = planToStreamColMap[ColIdx]` in each `aggExpr`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 0 of 6 files reviewed at latest revision, 31 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 576 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaHVeH8Z6S7z-zJOrq:-KZPpNwba_DypXb3by5d:b-3sow49) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L576)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

By the way - why is 1-1 mapping necessary for grouping? Can't we simply do `ColIdx = planToStreamColMap[ColIdx]` in each `aggExpr`?
</blockquote></details>

i'm not sure i follow, what's the second `ColIdx` in `ColIdx = planToStreamColMap[ColIdx]`? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/distsql_physical_planner.go, line 576 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235#-KYaHVeH8Z6S7z-zJOrq:-KZRz2GfzbuOP_MvpaUb:buiszra) ([raw file](https://github.com/cockroachdb/cockroach/blob/49292b342fd048705efe9047e412fd688d709737/pkg/sql/distsql_physical_planner.go#L576)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

i'm not sure i follow, what's the second `ColIdx` in `ColIdx = planToStreamColMap[ColIdx]`? 
</blockquote></details>

I mean instead of making sure that we have a 1-1 mapping (and thus `planToStreamColMap[i] = i`) can't we support an arbitrary mapping by just remapping the columns in the aggregator spec?
```
for i := range aggregatorSpec.Exprs {
  aggregatorSpec.Exprs[i].ColIdx = p.planToStreamColMap[aggregatorSpec.Exprs[i].ColIdx]
}
```

In the `[k k k]` example, this translates to: couldn't the expressions all reference `ColIdx=0` instead of us actually creating three columns with the value `k`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12235)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
I see. That may be possible yes, so for your original question of why I hadn't done this I hadn't thought of it then. Could be a viable alternative to introducing remapping logic into the spec.

RaduBerinde commented:
Cool, I was asking more to double-check if I'm missing something. Thank you!
